### ISSUE TITLE
Any interest in pulling out the `diff` machinery into its own package?

### ISSUE THREAD

**Issue by mooreryan** [giacomocavalieri/birdie#13]
Would you have any interest in pulling out the diff machinery into its own package?  Currently, I'm just copy/pasting the `diff` module plus some of the pretty printing stuff from the `birdie` module and using it in the projects that need it.  However, it would be nice if I could simply depend on a `diff` package instead.

There are a few parts of the diff stuff that are specific to birdie, but it didn't take too many code changes for me to be able to use it outside of birdie as well.  Then you could imagine writing just a bit of glue code to use the newly factored out `diff` package from birdie.

If this is something you would be interested in, I could make a proof of concept of it to demonstrate what I mean.  Though, if it is something you are not interested in, I can keep vendoring the parts I need--it isn't too bad to do it :)

**Comment by giacomocavalieri** [comment]
Hello! What would you need this for? It's not out of the table, but the implementation needs a bit of polish so I'm not super sure about releasing it in a standalone package for now. Also there's the `gap` package to get the diff of strings and lists and I'd like to first see if I could contribute to that rather than releasing my own package just for this

**Comment by mooreryan** [comment]
> What would you need this for?

Mainly for diffing the result of tests.  Eg this: https://github.com/maxdeviant/startest/issues/8, or if that doesn't go through, then directly implementing the diffs in the [qcheck](https://github.com/mooreryan/gleam_qcheck) library.

No worries if it's out-of-scope, I can vendor the stuff I need, too.

**Comment by mooreryan** [comment]
> The best way to show test diffs should work on the constructors themselves I think

Could explain what you mean?  I get what you mean about needed a good formatting for the values, but I do get decent result by using the pprint library prior to diffing, then following that with the histogram diff algorithm.

**Comment by giacomocavalieri** [comment]
Yeah the result is ok because pprint already does a great job in displaying the types in a diff-friendly way but I think we can make a way better job if the diff could be aware of the actual values it's diffing instead of working on strings directly.

Imagine you have this diff:
```diff
  Error(#(
    [
      "A list of values that doesn't change",
      "and is the same",
    ],
-   Ok(4),
-   "This changed",
-   ["Hello"],
+   Ok(3),
+   "Hello",
+   [],
  ))
```
With something aware of the values you could get semantic diffs instead, the imagination is the limit:
![carbon](https://github.com/giacomocavalieri/birdie/assets/20598369/411043cf-8dd1-4146-a1f0-219ac54609b0)

**Comment by mooreryan** [comment]
Oh okay, I see what you mean, thanks. (E.g., something like this: https://github.com/lambdaisland/deep-diff2, ofc that's clojure, but still.)

Yes, that would be better for sure. I will settle for the pprint+diff for now at least.
